lib/ansible/playbook/handler.py

```diff
@@ -59,7 +60,12 @@ def notify_host(self, host):
         return False
 
     def remove_host(self, host):
-        self.notified_hosts = [h for h in self.notified_hosts if h != host]
+        try:
+            self.notified_hosts.remove(host)
```

Comment:
Just curious about whether we could use a different data structure instead of a list.  Do we need to maintain order?  What are the expectations of code that may utilize this if we used a different data structure.  A `set` would be even faster, but if we needed order a `dict` could work, but there are obviously behavioral differences by switching to a different format. 

I do agree though that this is as fast as it could be with a list.

Reply:
I believe that the most used operation we use here is checking for whether a host is present (in `Handler.is_host_notified`, which is also used in `Handler.notify_host`) so I did experiment with using a `dict` instead of a `list` because of `dict`'s `k in d` performance but it felt weird adding a host via something like `Handler.notified_hosts[host.name] = ...`.

The only expectation is that the elements in the data structure are unique (so we don't run a handler more than once on a host). I'll play with replacing a `list` with a `set` (which I didn't think was possible before for now unknown reason 🤷🏻‍♂️ ).

Other concern I had with changing the data structure is that `Handler.notified_hosts` is somewhat "marked" public so I ended up creating a `notified_hosts` `@property` for backwards compatibility and then in the end I just changed the list comprehension to using `remove` for a minimal change :-)

Reply:
```diff
diff --git a/lib/ansible/playbook/handler.py b/lib/ansible/playbook/handler.py
index c9cba6a6c31..374f81a65ba 100644
--- a/lib/ansible/playbook/handler.py
+++ b/lib/ansible/playbook/handler.py
@@ -17,17 +17,24 @@
 
 from __future__ import annotations
 
+import typing as t
+
 from ansible.playbook.attribute import NonInheritableFieldAttribute
 from ansible.playbook.task import Task
 from ansible.module_utils.six import string_types
 
 
+if t.TYPE_CHECKING:
+    from ansible.inventory.host import Host
+
+
 class Handler(Task):
 
     listen = NonInheritableFieldAttribute(isa='list', default=list, listof=string_types, static=True)
 
     def __init__(self, block=None, role=None, task_include=None):
-        self.notified_hosts = []
+        # TODO backwards compat, do we care?
+        self.notified_hosts: set[Host] = set()
 
         self.cached_name = False
 
@@ -52,19 +59,19 @@ class Handler(Task):
         t = Handler(block=block, role=role, task_include=task_include)
         return t.load_data(data, variable_manager=variable_manager, loader=loader)
 
-    def notify_host(self, host):
-        if not self.is_host_notified(host):
-            self.notified_hosts.append(host)
-            return True
-        return False
+    def notify_host(self, host: Host):
+        len_before = len(self.notified_hosts)
+        self.notified_hosts.add(host)
+        len_after = len(self.notified_hosts)
+        return len_before != len_after
 
-    def remove_host(self, host):
-        self.notified_hosts = [h for h in self.notified_hosts if h != host]
+    def remove_host(self, host: Host):
+        self.notified_hosts.remove(host)
 
     def clear_hosts(self):
-        self.notified_hosts = []
+        self.notified_hosts.clear()
 
-    def is_host_notified(self, host):
+    def is_host_notified(self, host: Host):
         return host in self.notified_hosts
 
     def serialize(self):
```